llvm/lib/Analysis/CtxProfAnalysis.cpp

```diff
@@ -0,0 +1,91 @@
+//===- CtxProfAnalysis.cpp - contextual profile analysis ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of the contextual profile analysis, which maintains contextual
+// profiling info through IPO passes.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Analysis/CtxProfAnalysis.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/IR/Analysis.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/ProfileData/PGOCtxProfReader.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+namespace llvm {
+namespace json {
+Value toJSON(const PGOContextualProfile &P) {
+  Object Ret;
+  Ret["Guid"] = P.guid();
+  Ret["Counters"] = Array(P.counters());
+  auto AllCS =
+      ::llvm::map_range(P.callsites(), [](const auto &P) { return P.first; });
+  auto MaxIt = ::llvm::max_element(AllCS);
+  if (MaxIt != AllCS.end()) {
+    Array CSites;
+    // Iterate to, and including, the maximum index.
+    for (auto I = 0U; I <= *MaxIt; ++I) {
+      CSites.push_back(Array());
+      Array &Targets = *CSites.back().getAsArray();
+      if (P.hasCallsite(I))
+        for (const auto &[_, Ctx] : P.callsite(I))
+          Targets.push_back(toJSON(Ctx));
+    }
+    Ret["Callsites"] = std::move(CSites);
+  }
+  return Ret;
+}
+
+Value toJSON(const PGOContextualProfile::CallTargetMapTy &P) {
+  Array Ret;
+  for (const auto &[_, Ctx] : P)
+    Ret.push_back(toJSON(Ctx));
+  return Ret;
+}
+} // namespace json
+} // namespace llvm
+
+using namespace llvm;
+#define DEBUG_TYPE "ctx_prof"
```

Comment:
I think this can go after the includes, it seems a bit strange to see DEBUG_TYPE defined down here.